library/lua/dfhack.lua

```diff
@@ -163,6 +163,7 @@ COMMA = ","
 PERIOD = "."
 
 function printall(table)
+    if type(table) ~= 'table' then print(tostring(table)) return end
```

Comment:
This unfortunately breaks for userdata: `~scr` now just prints `<viewscreen_titlest: 0x260c000>` instead of also printing fields. `pcall(pairs, table)` below was intended to allow any type where `pairs()` works.

I think https://github.com/DFHack/dfhack/commit/e311e002c7890ad4340cca343ac74ad9a9035cd1#diff-44c92af21289640b1608413cfd1aef7aa3403eda7986ba7955e7a35bf32d72c1L212 is likely what changed the behavior here - `pairs()` no longer throws an error on invalid objects, like `nil`, but `next()` does.

Reply:
reworked to use a `safe_iterate()` function, which should allow all possible iterables to work.

`printall` and `printall_ipairs` now output a straight `tostring` of their input if iteration fails, but this has an undesirable effect when running, for example, `~('hi')`, which now prints out:
```
hi
hi
```
once by `print(table.unpack(data,2,data.n))` and once by `printall(data[2])`. What are your thoughts on how this should be handled? should `printall` go back to not printing anything for non-iterables?

Reply:
I was thinking something more along the lines of a drop-in replacement for `pairs()` would work, rather than something taking a callback, such as:
```lua
for k, v in safe_pairs(t) do
   ...
end
```

I suppose something like this could also be an option, although I can't think of use-cases other than `pairs()` and `ipairs()` off the top of my head:
```
for k, v in safe_iterate(pairs, t) do
   ...
end
```

I've found Lua iterators somewhat tricky to work with in the past, though, so your approach is easier to deal with. I'll see if I can wrangle something more similar to `pairs()` together.

Reply:
If we write a drop-in replacement for pairs that just never fails, then how can we tell if something is non-iterable? Or is that the desired behavior - just do nothing instead of failing?

Reply:
Yep, `printall()` and `~` have historically silently printed nothing for non-iterable types. The couple other uses of `pcall(pairs)` also seem to be trying to iterate only if a type is iterable and doing nothing otherwise.

Reply:
For reference, here's a very basic implementation of what I had in mind that worked in Lua 5.3.3:
```lua
function safe_pairs(t)
    if (pcall(pairs, t)) then
        return pairs(t)
    else
        return function() end
    end
end
```

I haven't gotten this to work with 5.3.6 quite yet (the issue is that pairs() returns `next` as its first return value, and _that_ is what errors in 5.3.6)

Reply:
Ok, here's something I came up with that works as I was hoping in 5.3.6:
```lua
function _wrap_iterator(next_fn, ...)
    local wrapped_iter = function(...)
        local ret = {pcall(next_fn, ...)}
        local ok = table.remove(ret, 1)
        if ok then
            return table.unpack(ret)
        end
    end
    return wrapped_iter, ...
end

function safe_pairs(t)
    if (pcall(pairs, t)) then
        return _wrap_iterator(pairs(t))
    else
        return function() end
    end
end
```

`safe_pairs()` could be parameterized to take in `pairs` or `ipairs`, of course.

Example use:
```lua
function safe_printall(t)
    for k, v in safe_pairs(t) do
        print(('%s = %s'):format(k, v))
    end
end
```

I have somewhat mixed feelings about _how much_ longer it is than before. There's also some overhead from wrapping every call to `next_fn` (aka `next()`) with `pcall()`, although your implementation also has overhead. Unfortunately, looking at [this section of the Lua manual](https://www.lua.org/manual/5.3/manual.html#3.3.5), the iterator function itself must remain constant between iterations (`next` is equivalent to `f` in their example).

Reply:
I see the utility of `safe_pairs` as defined above for use in scripts. It is a good API-compatible drop-in replacement for `pairs`. However, I'm not sure it's better than the current implementation for use within `dfhack.lua`. First, I'm not sure exactly how much overhead `pcall` adds, but the current PR's `safe_iterate()` function calls `pcall()` once and `safe_pairs()` above calls `pcall()` O(N) times. That's a theoretical argument, though. I'll get some hard data on whether this matters when I can.

Second, though, regardless of whether we add `safe_pairs()` to `dfhack.lua` (and I think we should) we would still need a function like the current `safe_iterate()` for use in the `print_fields()` method, which needs to know *whether* an object was successfully iterated so it can report if iteration failed. And if we need the current, more processing-efficient implementation anyway, why not use it as the non-user-visible backend to `printall()`?

Reply:
>I have somewhat mixed feelings about how much longer it is than before 

I can help with that:
```lua
function safe_next(t, k)
    local ok, next_k, v = pcall(next, t, k)
    if ok then return next_k, v
    else return nil
    end
end

function safe_pairs (t)
    return safe_next, t, nil
end
```

Reply:
this is indeed shorter, but where does it actually call `pairs(t)`?

Reply:
> this is indeed shorter, but where does it actually call `pairs(t)`?

Yeah, forgot about __pairs. How about this one then:
```lua
function safe_pairs (t)
    local ok, iter, t = pcall(pairs, t)
    function safe_iter(t, k)
        local ok, next_k, v = pcall(iter, t, k)
        if ok then return next_k, v
        else return nil
        end
    end
    return safe_iter, t, nil
end
```

Reply:
Now that I actually look at the code side by side, my version is basically slightly refactored `_wrap_iterator` from above without vararg support (is it needed?). That's a pity.

Reply:
Vararg support was mostly me taking a shortcut, since we don't do anything with the parameters other than passing them through. Per https://www.lua.org/manual/5.3/manual.html#3.3.5, `next()` (`iter` in your example) is only ever called with 2 parameters, but it can return any number of values, which is why I collected all return values from calling it in my implementation. `pairs()` and `ipairs()` only take one value, but another iterator could potentially take more, which is why I made `_wrap_iterator()` take varargs.